### PR TITLE
Caching inpainting data + some updates to the one degree simulation

### PR DESCRIPTION
**Pull request by glwagner**
The main thing this PR does is refactor `ECCO_field` so that inpainted data is cached and reloaded. The inpainted data is saved to a `jld2` file. This really speeds up prototyping especially with small models (such as the three degree model in `one_degree_simulation.jl`) where inpainting dominates the cost of initializing a simulation.

### CODE REVIEW

File: Project.toml
Comment (navidcy) [inline Project.toml:42]: does this exist?

Comment (navidcy) [reply Project.toml:42]: oh I see.. it wasn't registered

https://github.com/CliMA/OrthogonalSphericalShellGrids.jl/commit/5ad39ae766c292332f316289af2f0ef1c0919e55#commitcomment-148922634

Comment (glwagner) [reply Project.toml:42]: now we have to bump to 0.1.8

**Comment by glwagner** [comment]
Just to clarify, the bump to 0.1.8 is also required to use the latest Oceananigans

**Comment by navidcy** [comment]
What do you mean "latest": latest tagged or dev version?

**Comment by navidcy** [comment]
when on GPU, I get the following:

```Julia
julia> maximum(simulation.model.ocean.model.velocities.u)
ERROR: Number of threads per block exceeds kernel limit (896 > 768).
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] diagnose_launch_failure(f::CUDA.CuFunction, err::CUDA.CuError; blockdim::CUDA.CuDim3, threaddim::CUDA.CuDim3, shmem::Int64)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:120
  [3] launch(::CUDA.CuFunction, ::CUDA.KernelState, ::CartesianIndices{…}, ::CartesianIndices{…}, ::CUDA.CuDeviceArray{…}, ::Oceananigans.AbstractOperations.ConditionalOperation{…}; blocks::Int64, threads::Int64, cooperative::Bool, shmem::Int64, stream::CUDA.CuStream)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:73
  [4] launch
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:52 [inlined]
  [5] #972
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:189 [inlined]
  [6] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:149 [inlined]
  [7] macro expansion
    @ ./none:0 [inlined]
  [8] convert_arguments
    @ ./none:0 [inlined]
  [9] #cudacall#971
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:191 [inlined]
 [10] cudacall
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:187 [inlined]
 [11] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:279 [inlined]
 [12] macro expansion
    @ ./none:0 [inlined]
 [13] AbstractKernel
    @ ./none:0 [inlined]
 [14] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:114 [inlined]
 [15] mapreducedim!(f::typeof(identity), op::typeof(max), R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…}; init::Nothing)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/mapreduce.jl:271
 [16] mapreducedim!(f::typeof(identity), op::typeof(max), R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…})
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/mapreduce.jl:169
 [17] mapreducedim!(f::Function, op::Function, R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…})
    @ GPUArrays ~/.julia/packages/GPUArrays/qt4ax/src/host/mapreduce.jl:10
 [18] maximum!(f::Function, r::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…}; init::Bool)
    @ Base ./reducedim.jl:1036
 [19] maximum!(f::Function, r::Field{…}, a::Oceananigans.AbstractOperations.ConditionalOperation{…}; condition::Nothing, mask::Float64, kwargs::@Kwargs{…})
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:670
 [20] maximum(f::Function, c::Field{…}; condition::Nothing, mask::Float64, dims::Function)
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:700
 [21] maximum
    @ ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:689 [inlined]
 [22] maximum(c::Field{…})
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:709
 [23] top-level scope
    @ REPL[62]:1

caused by: CUDA error: too many resources requested for launch (code 701, ERROR_LAUNCH_OUT_OF_RESOURCES)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/libcuda.jl:30
  [2] check
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/libcuda.jl:37 [inlined]
  [3] cuLaunchKernel
    @ ~/.julia/packages/CUDA/2kjXI/lib/utils/call.jl:34 [inlined]
  [4] (::CUDA.var"#966#967"{Bool, Int64, CUDA.CuStream, CUDA.CuFunction, CUDA.CuDim3, CUDA.CuDim3})(kernelParams::Vector{Ptr{Nothing}})
    @ CUDA ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:66
  [5] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:33 [inlined]
  [6] macro expansion
    @ ./none:0 [inlined]
  [7] pack_arguments(::CUDA.var"#966#967"{…}, ::CUDA.KernelState, ::CartesianIndices{…}, ::CartesianIndices{…}, ::CUDA.CuDeviceArray{…}, ::Oceananigans.AbstractOperations.ConditionalOperation{…})
    @ CUDA ./none:0
  [8] launch(::CUDA.CuFunction, ::CUDA.KernelState, ::CartesianIndices{…}, ::CartesianIndices{…}, ::CUDA.CuDeviceArray{…}, ::Oceananigans.AbstractOperations.ConditionalOperation{…}; blocks::Int64, threads::Int64, cooperative::Bool, shmem::Int64, stream::CUDA.CuStream)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:59
  [9] launch
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:52 [inlined]
 [10] #972
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:189 [inlined]
 [11] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:149 [inlined]
 [12] macro expansion
    @ ./none:0 [inlined]
 [13] convert_arguments
    @ ./none:0 [inlined]
 [14] #cudacall#971
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:191 [inlined]
 [15] cudacall
    @ ~/.julia/packages/CUDA/2kjXI/lib/cudadrv/execution.jl:187 [inlined]
 [16] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:279 [inlined]
 [17] macro expansion
    @ ./none:0 [inlined]
 [18] AbstractKernel
    @ ./none:0 [inlined]
 [19] macro expansion
    @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:114 [inlined]
 [20] mapreducedim!(f::typeof(identity), op::typeof(max), R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…}; init::Nothing)
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/mapreduce.jl:271
 [21] mapreducedim!(f::typeof(identity), op::typeof(max), R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…})
    @ CUDA ~/.julia/packages/CUDA/2kjXI/src/mapreduce.jl:169
 [22] mapreducedim!(f::Function, op::Function, R::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…})
    @ GPUArrays ~/.julia/packages/GPUArrays/qt4ax/src/host/mapreduce.jl:10
 [23] maximum!(f::Function, r::SubArray{…}, A::Oceananigans.AbstractOperations.ConditionalOperation{…}; init::Bool)
    @ Base ./reducedim.jl:1036
 [24] maximum!(f::Function, r::Field{…}, a::Oceananigans.AbstractOperations.ConditionalOperation{…}; condition::Nothing, mask::Float64, kwargs::@Kwargs{…})
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:670
 [25] maximum(f::Function, c::Field{…}; condition::Nothing, mask::Float64, dims::Function)
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:700
 [26] maximum
    @ ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:689 [inlined]
 [27] maximum(c::Field{…})
    @ Oceananigans.Fields ~/.julia/packages/Oceananigans/p4IUv/src/Fields/field.jl:709
 [28] top-level scope
    @ REPL[62]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by glwagner** [comment]
Ok, this is a problem with reductions on immersed boundary grids unfortunately. For the moment, we need to use `maximum(interior(u))`.

**Review by navidcy — APPROVED** [review]

File: experiments/one_degree_simulation/one_degree_simulation.jl
Comment (navidcy) [inline experiments/one_degree_simulation/one_degree_simulation.jl:1]: it worths renaming the file if we are gonna leave it as 3 degree horizontal resolution

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
> Ok, this is a problem with reductions on immersed boundary grids unfortunately. For the moment, we need to use `maximum(interior(u))`.

I changed that in the progress function

**Comment by glwagner** [comment]
> What do you mean "latest": latest tagged or dev version?

The latest tagged

**Comment by navidcy** [comment]
Is this ready?

**Comment by navidcy** [comment]
(if so, can we not merge until the docs built on main finishes to see if docs get deployed?)

**Comment by glwagner** [comment]
This PR needs tests and the docstring for `ECCO_field` is still unfinished.

**Comment by glwagner** [comment]
> Is this ready?

It can be merged if tests pass, but it would be nice to add a test that ensures the inpainting caching is working.

**Comment by glwagner** [comment]
@navidcy ok I think this is ready. I decided that `ECCO_field` is not user facing so we don't need a groomed docstring.

**Comment by navidcy** [comment]
some tests were broken; could you have a look @glwagner?

**Comment by simone-silvestri** [comment]
I think it would be nice to add a keyword argument to make the caching optional.

**Comment by simone-silvestri** [comment]
also, the ECCO_field function should use `@root begin` for writing down the unpainted data to make sure there are no conflicts between different processes

**Comment by simone-silvestri** [comment]
Actually, I think it is necessary to give an option to toggle the writing / reading because we might want to use a different inpainting procedure than the one of the saved file.

Another option is to add some metadata with the inpainting steps used to the file and read it in only if it matches with the requested `ECCO_field`

**Comment by glwagner** [comment]
> Actually, I think it is necessary to give an option to toggle the writing / reading because we might want to use a different inpainting procedure than the one of the saved file.
> 
> Another option is to add some metadata with the inpainting steps used to the file and read it in only if it matches with the requested `ECCO_field`

That would be nice, though it can be achieved currently by deleting the inpainting files...

**Comment by glwagner** [comment]
It doesn't m,atter if this works on GPU

**Comment by glwagner** [comment]
Could this PR be the issue? @navidcy @simone-silvestri

**Pull request merged**
